[MRG] Add swiss roll with hole dataset

pujaltes opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

pujaltes committed:
Add swiss roll with hole dataset

Added the option for creating the swiss roll with hole dataset in the make_swiss_roll() function.

pujaltes committed:
Added swiss roll with hole dataset

glemaitre commented:
is there any related issue related to this enhancement?

pujaltes commented:
@glemaitre No, I have not seen an issue related to this. 

I'm working on comparing various non-linear dimensionality reduction algorithms and wanted to use the swiss roll and swiss roll with hole data. The swiss roll with hole is a classic dataset for evaluating non-linear dimensionality reduction and I felt that it's important to include the option to create it.

pujaltes committed:
Update _samples_generator.py

pujaltes committed:
Merge branch 'scikit-learn:main' into patch-1

ogrisel commented:
Could you please update/extend one of the existing examples to demonstrate how manifold learning models of scikit-learn behave on this dataset?

For instance, the following example could be extended into a multi-stage example (first without the hole and then with the hole):

https://scikit-learn.org/stable/auto_examples/manifold/plot_swissroll.html#sphx-glr-auto-examples-manifold-plot-swissroll-py

This example dates from a long time ago, before sphinx-gallery had support for notebook style examples. Nowadays we would write multi-step examples with the `# %%` cell separation marker. For instance look at for an example with a modern layout: https://scikit-learn.org/dev/auto_examples/applications/plot_cyclical_feature_engineering.html

pujaltes commented:
@ogrisel Sure, I'll give it a try and update when done.

pujaltes committed:
Merge branch 'scikit-learn:main' into patch-1

pujaltes committed:
Revamped swiss roll example

Updated Swiss Roll example to include Swiss-hole and improved the general look of the example.

pujaltes commented:
@ogrisel Sorry for the delay. 

I have updated the example as requested and also included a comparison to t-SNE. Let me know if there's anything else I can do.

pujaltes committed:
Black formatting

pujaltes committed:
More black linting

pujaltes committed:
Merge branch 'scikit-learn:main' into patch-1

pujaltes committed:
pep-8

pujaltes committed:
fixed doc for CircleCi

For some reason, CircleCi did gave the error: "AttributeError: Unknown property auto_add_to_figure", so I changed the method of creating the 3d plot.

pujaltes committed:
removed unused import

pujaltes committed:
Merge branch 'scikit-learn:main' into patch-1

pujaltes committed:
Merge branch 'scikit-learn:main' into patch-1
